pkg/reconciler/revision/resources/deploy.go

```diff
@@ -136,49 +135,60 @@ func makePodSpec(rev *v1.Revision, loggingConfig *logging.Config, tracingConfig
 	return podSpec, nil
 }
 
-// BuildUserContainer makes a container from the Revision template.
-func BuildUserContainer(rev *v1.Revision) *corev1.Container {
-	userContainer := rev.Spec.GetContainer().DeepCopy()
+// BuildUserContainers makes a container from the Revision template.
+func BuildUserContainers(rev *v1.Revision) []corev1.Container {
+	containers := make([]corev1.Container, 0, len(rev.Spec.PodSpec.Containers))
+	for i := range rev.Spec.PodSpec.Containers {
+		var container corev1.Container
+		if len(rev.Spec.PodSpec.Containers[i].Ports) != 0 || len(rev.Spec.PodSpec.Containers) == 1 {
+			container = makeServingContainer(rev.Spec.PodSpec.Containers[i], rev)
+		} else {
+			container = makeContainer(rev.Spec.PodSpec.Containers[i], rev)
+		}
+		if rev.Status.ContainerStatuses != nil {
+			container.Image = rev.Status.ContainerStatuses[i].ImageDigest
```

Comment:
Is ContainerStatuses always guaranteed to have len(Containers) here? We’ll panic and crash the whole controller if not..

Reply:
Yes its guaranteed

But there is a situation here https://github.com/knative/serving/blob/master/pkg/webhook/podspec_dryrun.go#L60 where its calling `BuildUserContainers` function by passing just `spec` 
To avoid panic for that added condition to check whether `ContainerStatuses` is nil or not otherwise it will panic because of this line `rev.Status.ContainerStatuses[i].ImageDigest`

Reply:
yeah, looks like apart from dry run we always resolve digests before this is called so I think it's fine. I still worry a bit that this means that we treat `rev.Status.ContainerStatuses == nil` and `len(rev.Status.ContainerStatuses) == 0` differently, which [we'd ideally avoid](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices) though (especially since the len(0) case actually panics).

Might also be worth adding a bit of documentation that this method will use `status.ContainerStatuses`, if present, to populate the container images (since it's at least slightly unusual for it to be doing that, and people may not realise that calling with status out of sync or before image digests are resolved could cause problems)?

Reply:
One other thing - I can't figure out how this line works if ImageDigest isn't set? e.g. if we're skipping the digest resolution (for example in the ko.local case) it looks like [`Resolve` returns `"", nil`](https://github.com/knative/serving/blob/e8fe5efc640acc6da99c545fc826abb2c997a207/pkg/reconciler/revision/resolve.go#L109) so won't `ContainerStatus[I].ImageDigest` - and therefore, because of this line `container.Image`-  be empty string?

Reply:
in that case `container.Image` will have the user provide image info not the digested value

Reply:
> One other thing - I can't figure out how this line works if ImageDigest isn't set? e.g. if we're skipping the digest resolution (for example in the ko.local case) it looks like [`Resolve` returns `"", nil`](https://github.com/knative/serving/blob/e8fe5efc640acc6da99c545fc826abb2c997a207/pkg/reconciler/revision/resolve.go#L109) so won't `ContainerStatus[I].ImageDigest` - and therefore, because of this line `container.Image`- be empty string?

in that case container.Image will have the user provide image info not the digested value

Reply:
I'm not sure I understand where that happens? If resolve returns `"", nil` (which it does if the image is ko.local) then ImageDigest here is empty string. Is there somewhere defaulting it back to container.Image that I missed?

Reply:
FWIW I checked out the branch and tried manually adding a sidecar with a ko.local image and got the error below in the controller logs, indicating we tried to create a deployment with an empty container.Image:

~~~~
{"level":"info","ts":"2020-05-18T11:38:43.391Z","logger":"controller.revision-controller.event-broadcaster","caller":"record/event.go:274","msg":"Event(v1.ObjectReference{Kind:\"Revision\", Namespace:\"default\", Name:\"helloworld-go-75p2l\", UID:\"1dc5d2be-817d-451c-8484-80ee167b76d3\", APIVersion:\"serving.knative.dev/v1\", ResourceVersion:\"441660\", FieldPath:\"\"}): type: 'Warning' reason: 'InternalError' failed to create deployment \"helloworld-go-75p2l-deployment\": Deployment.apps \"helloworld-go-75p2l-deployment\" is invalid: spec.template.spec.containers[1].image: Required value","commit":"9bda154","knative.dev/controller":"revision-controller"}
~~~~

here's the ksvc I used:
~~~~
apiVersion: serving.knative.dev/v1 # Current version of Knative
kind: Service
metadata:
  name: helloworld-go
  namespace: default
spec:
  template:
    spec:
      containers:
        - image: gcr.io/knative-samples/helloworld-go
          ports:
            - containerPort: 80
          env:
            - name: TARGET
              value: "Go Sample v1"
        - image: ko.local/foo/bar # <-- this is the ko.local sidecar that ends up empty
~~~~

Reply:
> One other thing - I can't figure out how this line works if ImageDigest isn't set? e.g. if we're skipping the digest resolution (for example in the ko.local case) it looks like [`Resolve` returns `"", nil`](https://github.com/knative/serving/blob/e8fe5efc640acc6da99c545fc826abb2c997a207/pkg/reconciler/revision/resolve.go#L109) so won't `ContainerStatus[I].ImageDigest` - and therefore, because of this line `container.Image`- be empty string?

Right with the current code if there is no digest value image will be replaced with `" "` so i need to handle that scenario same like [earlier](https://github.com/knative/serving/blob/master/pkg/reconciler/revision/resources/deploy.go#L163-L165)

Reply:
(that's odd, did you delete a comment in this thread or has GitHub gone weird? I could swear I was replying to something above?)

Reply:
> (that's odd, did you delete a comment in this thread or has GitHub gone weird? I could swear I was replying to something in [#7801 (comment)](https://github.com/knative/serving/pull/7801#discussion_r426562582) but now the message I was replying to has disappeared?)

Yes initially added some comment but thats seems to be wrong so i deleted and added the right comment

Reply:
But i replied before this [comment](https://github.com/knative/serving/pull/7801/files#r426562582) :thinking: it didn't showed me your latest comments itself

Reply:
🤔. 

Anyway: main point is I think we need to handle this case, be resilient to an empty rather than nil array, and document (since it's fairly easy not to realise) that this method will use any resolved digests in status, if non-empty, to override the container images.

Reply:
@julz updated code 
Please have a look 
Thank you